### PR TITLE
[DO NOT MERGE] Feat/alternative v4 rfq implementation [WIP]

### PR DESCRIPTION
**Pull request by alexkroeger**
## Description

Added an alternative RFQ implementation to asset-swapper.

## Testing instructions

Extended quote-requestor tests to cover alternative implementation.

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                        .decimalPlaces(altPair.baseAssetDecimals, BigNumber.ROUND_DOWN),
+                    altPair.baseAssetDecimals,
+                );
+            } else {
+                throw new Error('niether amount or value were specified');
+            }
+        } else if (makerToken === altPair.quoteAsset) {
+            if (altIndicativeQuoteReponse.amount) {
+                // base asset = taker asset, quote asset = maker asset, amount specified
+                takerAmount = Web3Wrapper.toBaseUnitAmount(
```
Comment (PirosB3) [inline packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:83]: Is there an opportunity to refactor these chunks of code that look fairly similar?

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    altPair: AltOffering,
+    makerToken: string,
+    takerToken: string,
+): V4RFQIndicativeQuote {
+    let makerAmount: BigNumber;
+    let takerAmount: BigNumber;
+
+    if (!altIndicativeQuoteReponse.price) {
+        throw new Error('Price not returned by alt MM');
+    } else {
```
Comment (PirosB3) [inline packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:51]: We can drop the `else` here right?

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    return {
+        makerToken,
+        makerAmount,
+        takerToken,
+        takerAmount,
+        // HACK: alt implementation does not return an expiration with indicative quotes
+        // return now + { IMPUTED EXPIRY SECONDS } to have it included after order checks
+        // tslint:disable-next-line:custom-no-magic-numbers
+        expiry: new BigNumber(
+            new BigNumber(Date.now() / 1000).integerValue(BigNumber.ROUND_DOWN).plus(IMPUTED_EXPIRY_SECONDS),
```
Comment (PirosB3) [inline packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:122]: You can replace 1000 with `ONE_SECOND`

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+    if (altPair) {
+        const side =
+            altPair.baseAsset === takerRequestQueryParams.buyTokenAddress ? AltQuoteSide.Buy : AltQuoteSide.Sell;
+
+        // comparison price needs to be quote/base
+        // in the standard implementation, it's maker/taker
+        const altComparisonPrice =
+            altPair.quoteAsset === makerToken
+                ? takerRequestQueryParams.comparisonPrice
```
Comment (PirosB3) [inline packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:154]: Wouldn't `takerRequestQueryParams.comparisonPrice` be already `undefined`? do we need that tenary statement?

File: packages/asset-swapper/src/swap_quoter.ts [hunk 5]
```diff
@@ -342,6 +342,8 @@ export class SwapQuoter {
         if (calcOpts.rfqt !== undefined) {
             calcOpts.rfqt.quoteRequestor = new QuoteRequestor(
                 rfqtOptions ? rfqtOptions.makerAssetOfferings || {} : {},
+                rfqtOptions ? rfqtOptions.altRfqApiKey || '' : '',
```
Comment (PirosB3) [inline packages/asset-swapper/src/swap_quoter.ts:345]: Have we considered using `undefined` instead of an empty string?

**Review by PirosB3 — CHANGES_REQUESTED** [review]
There is quite a bit of logic here that I have to really digest, while I'm still understaning the code at a deeper level I've started adding some feedback

**Review by PirosB3 — CHANGES_REQUESTED** [review]
There is quite a bit of logic here that I have to really digest, while I'm still understaning the code at a deeper level I've started adding some feedback

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts (hunk 2 above)
Comment (alexkroeger) [reply packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:51]: price is used below so we need to establish that it's defined

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    altIndicativeQuoteReponse: AltIndicativeQuoteReponse,
+    altPair: AltOffering,
+    makerToken: string,
+    takerToken: string,
+): V4RFQIndicativeQuote {
+    let makerAmount: BigNumber;
+    let takerAmount: BigNumber;
+
+    if (!altIndicativeQuoteReponse.price) {
+        throw new Error('Price not returned by alt MM');
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:50]: quoteRequestor currently catches all errors that would occur here. we should discuss what we consider an error worth logging tho.

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        ) {
+            return offering;
+        }
+    }
+    return undefined;
+}
+
+function parseFirmQuoteResponseFromAltMM(altFirmQuoteReponse: AltFirmQuoteResponse): V4RFQFirmQuote {
+    return {
+        signedOrder: altFirmQuoteReponse.data['0xv4order'],
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:36]: it should be present if the quote is not 'rejected'

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+const IMPUTED_EXPIRY_SECONDS = 120;
+
+function getAltMarketInfo(
+    offerings: AltOffering[],
+    buyTokenAddress: string,
+    sellTokenAddress: string,
+): AltOffering | undefined {
+    for (const offering of offerings) {
+        if (
+            (buyTokenAddress === offering.baseAsset && sellTokenAddress === offering.quoteAsset) ||
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:25]: I'll lowercase everything to be safe

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+                    altPair.quoteAssetDecimals,
+                );
+                makerAmount = Web3Wrapper.toBaseUnitAmount(
+                    new BigNumber(altIndicativeQuoteReponse.value)
+                        .dividedBy(new BigNumber(altIndicativeQuoteReponse.price))
+                        .decimalPlaces(altPair.baseAssetDecimals, BigNumber.ROUND_DOWN),
+                    altPair.baseAssetDecimals,
+                );
+            } else {
+                throw new Error('niether amount or value were specified');
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:78]: it depends on if the quantity is specified in the base asset (`amount`), or the quote asset (`value`). so buys and sells can both be either/or

File: packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts (hunk 4 above)
Comment (alexkroeger) [reply packages/asset-swapper/src/utils/alt_mm_implementation_utils.ts:158]: don't understand the comment

File: packages/asset-swapper/src/constants.ts [hunk 10]
```diff
@@ -48,6 +48,8 @@ const DEFAULT_SWAP_QUOTER_OPTS: SwapQuoterOpts = {
     samplerGasLimit: 250e6,
     ethGasStationUrl: ETH_GAS_STATION_API_URL,
     rfqt: {
+        altRfqProfile: '',
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply packages/asset-swapper/src/constants.ts:51]: agreed that undefined is best, I'll refactor

**Pull request closed without merging**